edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/ReportedPropertiesValidator.cs

```diff
@@ -21,42 +21,44 @@ public void Validate(TwinCollection reportedProperties)
         {
             Preconditions.CheckNotNull(reportedProperties, nameof(reportedProperties));
             JToken reportedPropertiesJToken = JToken.Parse(reportedProperties.ToJson());
-            ValidateTwinProperties(reportedPropertiesJToken, 1);
             ValidateTwinCollectionSize(reportedProperties);
-            ValidateArrayContent(reportedPropertiesJToken);
+            // root level has no property name.
+            ValidateToken(string.Empty, reportedPropertiesJToken, 0, false);
         }
 
-        static void ValidateTwinProperties(JToken properties, int currentDepth)
+        static void ValidateToken(string name, JToken item, int currentDepth, bool inArray)
         {
-            foreach (JProperty kvp in ((JObject)properties).Properties())
-            {
-                ValidatePropertyNameAndLength(kvp.Name);
-
-                ValidateValueType(kvp.Name, kvp.Value);
+            ValidatePropertyNameAndLength(name);
 
-                if (kvp.Value is JValue)
+            if (item is JObject @object)
+            {
+                if (currentDepth > TwinPropertyMaxDepth)
                 {
-                    if (kvp.Value.Type is JTokenType.Integer)
-                    {
-                        ValidateIntegerValue(kvp.Name, (long)kvp.Value);
-                    }
-                    else
-                    {
-                        string s = kvp.Value.ToString();
-                        ValidatePropertyValueLength(kvp.Name, s);
-                    }
+                    throw new InvalidOperationException($"Nested depth of twin property exceeds {TwinPropertyMaxDepth}");
                 }
 
-                if (kvp.Value != null && kvp.Value is JObject)
+                // do validation recursively
+                foreach (JProperty kvp in @object.Properties())
                 {
-                    if (currentDepth > TwinPropertyMaxDepth)
-                    {
-                        throw new InvalidOperationException($"Nested depth of twin property exceeds {TwinPropertyMaxDepth}");
-                    }
+                    ValidateToken(kvp.Name, kvp.Value, currentDepth + 1, inArray);
+                }
+            }
+
+            if (item is JValue value)
+            {
+                ValidateValueType(name, value);
+                ValidateValue(name, value, inArray);
+            }
 
-                    // do validation recursively
-                    ValidateTwinProperties(kvp.Value, currentDepth + 1);
+            if (item is JArray array)
+            {
+                if (currentDepth > TwinPropertyMaxDepth)
+                {
+                    throw new InvalidOperationException($"Nested depth of twin property exceeds {TwinPropertyMaxDepth}");
```

Comment:
Is this applicable to arrays as well? Have you confirmed with IoTHub behavior?

Reply:
If this is true for all json type fields then can we just move it out of the destructure ifs and put a single if depth check at the top of this fn?

Reply:
@nyanzebra, this is true for arrays and objects only. But you might have a point, so I moved that into a func to avoid duplications.

@varunpuranik, as discussed offline with the hub folks, I believe we need to keep nested arrays validaion.